### PR TITLE
[primTorch] add ref for `cumprod`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #98670

### CODE REVIEW

File: test/test_meta.py [hunk 1]
```diff
@@ -900,11 +900,13 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
 meta_dispatch_early_skips = set({
     torch.Tensor.float_power_,
     # Errors out in one of the tests, while ProxyTensor passes...
+    torch.Tensor.cumprod_,
```
Comment (nkaretnikov) [inline test/test_meta.py:903]: Due to `_reduction` (used by `sum`/`prod`). `out` and `dtype` disagree. Might leave this for later.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -3980,12 +3982,12 @@ def unfold_copy(self: TensorLikeType, dimension: int, size: int, step: int):
...
 
-@register_decomposition(aten.cumsum)
-def cumsum(
+def _cumsumprod_common(
+    func,
+    init,
     a: TensorLikeType,
     dim: int,
     *,
-    keepdim: bool = False,
```
Comment (nkaretnikov) [inline torch/_refs/__init__.py:3988]: Don't see `keepdim` in the torch/ATen signature, so removed it.

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -3994,14 +3996,36 @@ def cumsum(
...
-        return sum(a.unsqueeze(0), dim=0, keepdim=keepdim, dtype=dtype, out=out)
+        return func(a.unsqueeze(0), dim=0, dtype=dtype, out=out)
     a = a.unsqueeze(dim + 1)
     rg = torch.arange(a.shape[dim], device=a.device)
     mask = rg.unsqueeze(1) <= rg
     for _ in range(ndim - dim - 1):
         mask = mask.unsqueeze(-1)
-    masked_a = utils.mask_tensor(mask, a)
-    return sum(masked_a, dim=dim, keepdim=keepdim, dtype=dtype, out=out)
+    masked_a = torch.where(mask, a, init)
```
Comment (nkaretnikov) [inline torch/_refs/__init__.py:4005]: `mask_tensor` (used before) is `where` if dtype is not bool. I need to pass `init=1` to make `prod` work properly.

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -3365,7 +3365,6 @@ def inplace_op(*args, **kwargs):
 register_inplace(aten.addmm_, aten.addmm)
 register_inplace(aten.addmv_, aten.addmv)
 register_inplace(aten.baddbmm_, aten.baddbmm)
-register_inplace(aten.cumprod_, aten.cumprod)
```
Comment (nkaretnikov) [inline torch/_decomp/decompositions.py:3368]: Moved to the ref file, next to `cumsum`.

**Review by ezyang — APPROVED** [review]

**Comment by nkaretnikov** [comment]
@ngimel OK, I will!

**Pull request closed without merging**